### DISCUSSION TITLE
How to create and use a local NuGet package for Avalonia.

### DISCUSSION THREAD

**Discussion by ShrutiJaiswal1494** [AvaloniaUI/Avalonia#6215] [Q&A] [answered]
Hi,

I am trying to test the `feature/ui-automation` branch. For this, I need to create a local NuGet package for Avalonia and use it in my project. Based on the documentation [online ](https://www.codeproject.com/Articles/1214420/Creating-a-Nuget-Package-Step-by-Step)it seems like to create a nugget package we need to have a `.nuspec` file but I am not able to find any in the Avalonia repo. 

Is there a way to create a local NuGet package for Avalonia. Any help is much appreciated.
Thanks.

**Comment by maxkatz6** [comment, marked as answer]
Build docs: https://github.com/AvaloniaUI/Avalonia/blob/master/Documentation/build.md
To run full build with packages you need either run "_build.csproj" or run "build.ps1" (build.sh on linux) script.

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6  I was following the above documentation and was able to build the project. I am wondering how can I create a NuGet package from my own local version of Avalonia so that I can use it in my application.

**Reply by maxkatz6** [reply]
Script does create packages. Check "artifacts" folder.

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6 I am getting the below error while running the `build.ps1` script. Though I see the DLL file is there in the mentioned path `C:\Users\shjaiswal\source\repos\LatestAvalonia\Avalonia\tests\Avalonia.LeakTests\bin\Release\net461\Avalonia.LeakTests.dll`. Is there a way to skip the tests or any suggestions on how to resolve this?

Thanks in advance!

![image](https://user-images.githubusercontent.com/79670853/125007229-a3fa7500-e014-11eb-88d7-ca350b3daf26.png)

**Reply by maxkatz6** [reply]
Leak tests might randomly fail, and run successfully from second try.
But you can disable them with this param https://github.com/AvaloniaUI/Avalonia/blob/master/nukebuild/BuildParameters.cs#L16

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6  The script did not generate a `.nupkg` file for `Avalonia.Native` and this package is required by `Avalonia.Desktop`. Looking at the script run output seems like the `CompileNative` step was skipped. I am wondering if this task is the one responsible for generating the `Avalonia.Native` .nupkg file. How can I generate the .nupkg file for `Avalonia.Native`?

Thanks in advance!

![image](https://user-images.githubusercontent.com/79670853/125011432-50405980-e01d-11eb-8208-0f8d6c39a5c4.png)

**Reply by maxkatz6** [reply]
You can build macOS binaries only on macOS.
https://github.com/AvaloniaUI/Avalonia/blob/master/nukebuild/Build.cs#L140

I.e. to create all packages, you will need to run pipeline on both windows and macos.
https://github.com/AvaloniaUI/Avalonia/blob/master/Documentation/build.md#additional-requirements-for-macos

**Reply by maxkatz6** [reply]
I don't know actually if you can use official Avalonia.Native build with your modified Avalonia.
CC @danwalmsley

**Reply by kekekeks** [reply]
They are compatible as long as native headers are unchanged

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6 @kekekeks I am trying to test accessibility support for Windows by pulling[ this PR locally](https://github.com/AvaloniaUI/Avalonia/pull/5177/files). I am not sure if using the standard version of `Avalonia.Native` will do the job since the PR has some changes in `Avalonia.Native` files as well.

**Reply by maxkatz6** [reply]
If you need it only for windows, you don't need Avalonia.Native, it's only for macos. 
Also to test you could use submodules

**Reply by ShrutiJaiswal1494** [reply]
@maxkatz6 @kekekeks Is there a way to force `Avalonia.Desktop` to use my local version while `Avalonia.Native` will use the standard released version.? I am trying to update the packages version in VS for `Abvalonia.Desktop` and it's looking for the latest local version for `Avaloni.Native` as well. 
Thanks in advance!

![image](https://user-images.githubusercontent.com/79670853/125131233-3e12f980-e0b7-11eb-8df6-700c8a311bee.png)

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/7c1a61bb7eb466f0bfb62dc407a823e318667b84/src/Avalonia.Native/Avalonia.Native.csproj#L5-L6 Replace those lines with `<IsPackable>true</IsPackable>` and remove lines here - https://github.com/AvaloniaUI/Avalonia/blob/7c1a61bb7eb466f0bfb62dc407a823e318667b84/src/Avalonia.Native/Avalonia.Native.csproj#L11-L18

Then build your nuget packages. That should produce Avalonia.Package package without native bits.

**Reply by ShrutiJaiswal1494** [reply]
@kekekeks After making the above changes I am getting multiple errors on build. These files are the auto-generated ones. Do you have any suggestions on how to resolve these errors?
Thanks in advance!

![image](https://user-images.githubusercontent.com/79670853/125149702-530b7f00-e0ef-11eb-906a-11b8521b3627.png)

**Reply by ShrutiJaiswal1494** [reply]
Deleting the `obj` and `bin` folder worked
